### ISSUE TITLE
C++ std lib version mismatch errors when compiling hipSYCL

### ISSUE THREAD

**Issue by WageeshaR** [AdaptiveCpp/AdaptiveCpp#772]
Hi @illuhad
Now I'm compiling hipSYCL with Clang in a Arm platform ([before I did in a x86 one](https://github.com/illuhad/hipSYCL/issues/768)). Its default c++ library path has only version 7 libs (`/usr/include/c++/7`). However there are several other newer GCC versions I can load into the system as modules and I load gcc 9.

First when I try to `make install`, it says `fatal error: 'filesystem' file not found` and that is obvious because g++7 doesn't have stable filesystem lib. Now when I try to include the library path for c++9 lib in CMakeList.txt file and make again, it fails again with a bunch or other errors.

Below is the output when without including c++9 libs.
```
Scanning dependencies of target hipSYCL_clang
[  1%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/HipsyclClangPlugin.cpp.o
[  3%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/GlobalsPruningPass.cpp.o
[  5%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/LoopSplitterInlining.cpp.o
[  7%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/SplitterAnnotationAnalysis.cpp.o
[  9%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/IRUtils.cpp.o
[ 11%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/KernelFlattening.cpp.o
[ 13%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/LoopsParallelMarker.cpp.o
[ 15%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/PHIsToAllocas.cpp.o
[ 16%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/RemoveBarrierCalls.cpp.o
[ 18%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/CanonicalizeBarriers.cpp.o
[ 20%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/SimplifyKernel.cpp.o
[ 22%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/LoopSimplify.cpp.o
[ 24%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/PipelineBuilder.cpp.o
[ 26%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/SubCfgFormation.cpp.o
[ 28%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/UniformityAnalysis.cpp.o
[ 30%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/VectorShape.cpp.o
[ 32%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/VectorizationInfo.cpp.o
[ 33%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/AllocaSSA.cpp.o
[ 35%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/VectorShapeTransformer.cpp.o
[ 37%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/Region.cpp.o
[ 39%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/SyncDependenceAnalysis.cpp.o
[ 41%] Linking CXX shared library libhipSYCL_clang.so
[ 41%] Built target hipSYCL_clang
Scanning dependencies of target hipSYCL-rt
[ 43%] Building CXX object src/runtime/CMakeFiles/hipSYCL-rt.dir/application.cpp.o
[ 45%] Building CXX object src/runtime/CMakeFiles/hipSYCL-rt.dir/runtime.cpp.o
[ 47%] Building CXX object src/runtime/CMakeFiles/hipSYCL-rt.dir/error.cpp.o
[ 49%] Building CXX object src/runtime/CMakeFiles/hipSYCL-rt.dir/backend.cpp.o
[ 50%] Building CXX object src/runtime/CMakeFiles/hipSYCL-rt.dir/backend_loader.cpp.o
/home/ri-wshilpage/sycl-compilers/hipSYCL/src/runtime/backend_loader.cpp:36:10: fatal error: 'filesystem' file not found
#include <filesystem>
         ^~~~~~~~~~~~
1 error generated.
make[2]: *** [src/runtime/CMakeFiles/hipSYCL-rt.dir/build.make:115: src/runtime/CMakeFiles/hipSYCL-rt.dir/backend_loader.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:286: src/runtime/CMakeFiles/hipSYCL-rt.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
Below is the output when I added `include_directories("/opt/gcc/9.3.0/snos/include/g++")` into CMakeList.txt and `make install`.
```
Scanning dependencies of target hipSYCL_clang
[  1%] Building CXX object src/compiler/CMakeFiles/hipSYCL_clang.dir/HipsyclClangPlugin.cpp.o
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/HipsyclClangPlugin.cpp:29:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/FrontendPlugin.hpp:31:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/Frontend.hpp:31:
In file included from /opt/gcc/9.3.0/snos/include/g++/algorithm:61:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_algobase.h:67:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_iterator.h:66:
/opt/gcc/9.3.0/snos/include/g++/bits/ptr_traits.h:140:14: error: unknown type name '_GLIBCXX20_CONSTEXPR'
      static _GLIBCXX20_CONSTEXPR pointer
             ^
/opt/gcc/9.3.0/snos/include/g++/bits/ptr_traits.h:140:42: error: expected ';' at end of declaration list
      static _GLIBCXX20_CONSTEXPR pointer
                                         ^
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/HipsyclClangPlugin.cpp:29:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/FrontendPlugin.hpp:31:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/Frontend.hpp:31:
In file included from /opt/gcc/9.3.0/snos/include/g++/algorithm:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_algo.h:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_tempbuf.h:60:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_construct.h:59:
In file included from /opt/gcc/9.3.0/snos/include/g++/new:40:
/opt/gcc/9.3.0/snos/include/g++/exception:101:3: error: unknown type name '_GLIBCXX17_DEPRECATED'
  _GLIBCXX17_DEPRECATED
  ^
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/HipsyclClangPlugin.cpp:29:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/FrontendPlugin.hpp:31:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/Frontend.hpp:31:
In file included from /opt/gcc/9.3.0/snos/include/g++/algorithm:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_algo.h:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_tempbuf.h:60:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_construct.h:59:
/opt/gcc/9.3.0/snos/include/g++/new:125:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new(std::size_t) _GLIBCXX_THROW (std::bad_alloc)
^
/opt/gcc/9.3.0/snos/include/g++/new:127:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new[](std::size_t) _GLIBCXX_THROW (std::bad_alloc)
^
/opt/gcc/9.3.0/snos/include/g++/new:139:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new(std::size_t, const std::nothrow_t&) _GLIBCXX_USE_NOEXCEPT
^
/opt/gcc/9.3.0/snos/include/g++/new:141:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new[](std::size_t, const std::nothrow_t&) _GLIBCXX_USE_NOEXCEPT
^
/opt/gcc/9.3.0/snos/include/g++/new:148:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new(std::size_t, std::align_val_t)
^
/opt/gcc/9.3.0/snos/include/g++/new:150:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new(std::size_t, std::align_val_t, const std::nothrow_t&)
^
/opt/gcc/9.3.0/snos/include/g++/new:156:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new[](std::size_t, std::align_val_t)
^
/opt/gcc/9.3.0/snos/include/g++/new:158:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD void* operator new[](std::size_t, std::align_val_t, const std::nothrow_t&)
^
/opt/gcc/9.3.0/snos/include/g++/new:173:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD inline void* operator new(std::size_t, void* __p) _GLIBCXX_USE_NOEXCEPT
^
/opt/gcc/9.3.0/snos/include/g++/new:174:10: error: cannot initialize return object of type 'int *' with an lvalue of type 'void *'
{ return __p; }
         ^~~
/opt/gcc/9.3.0/snos/include/g++/new:175:1: error: unknown type name '_GLIBCXX_NODISCARD'
_GLIBCXX_NODISCARD inline void* operator new[](std::size_t, void* __p) _GLIBCXX_USE_NOEXCEPT
^
/opt/gcc/9.3.0/snos/include/g++/new:176:10: error: cannot initialize return object of type 'int *' with an lvalue of type 'void *'
{ return __p; }
         ^~~
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/HipsyclClangPlugin.cpp:29:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/FrontendPlugin.hpp:31:
In file included from /home/ri-wshilpage/sycl-compilers/hipSYCL/src/compiler/../../include/hipSYCL/compiler/Frontend.hpp:31:
In file included from /opt/gcc/9.3.0/snos/include/g++/algorithm:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_algo.h:62:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_tempbuf.h:60:
In file included from /opt/gcc/9.3.0/snos/include/g++/bits/stl_construct.h:61:
In file included from /opt/gcc/9.3.0/snos/include/g++/ext/alloc_traits.h:36:
/opt/gcc/9.3.0/snos/include/g++/bits/alloc_traits.h:304:7: error: unknown type name '_GLIBCXX_NODISCARD'
      _GLIBCXX_NODISCARD static pointer
      ^
/opt/gcc/9.3.0/snos/include/g++/bits/alloc_traits.h:304:40: error: expected ';' at end of declaration list
      _GLIBCXX_NODISCARD static pointer
                                       ^
/opt/gcc/9.3.0/snos/include/g++/bits/alloc_traits.h:442:7: error: unknown type name '_GLIBCXX_NODISCARD'
      _GLIBCXX_NODISCARD static pointer
      ^
/opt/gcc/9.3.0/snos/include/g++/bits/alloc_traits.h:442:40: error: expected ';' at end of declaration list
      _GLIBCXX_NODISCARD static pointer
                                       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [src/compiler/CMakeFiles/hipSYCL_clang.dir/build.make:63: src/compiler/CMakeFiles/hipSYCL_clang.dir/HipsyclClangPlugin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:202: src/compiler/CMakeFiles/hipSYCL_clang.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```

I don't understand how Clang picks up standard header files and why there are errors when I included headers for version 9 (or any version that has stable filesystem lib). However, when I try compiling with G++ it compiles with no errors.

**Comment by fodinabor** [comment]
Hi, I have the feeling, you might be building hipSYCL on Isambard?
When using Isambard with hipSYCL, I ended up building everything manually, including boost and LLVM, not sure that's actually necessary.. I be missing crucial details, but I'll try to summarize how I used hipSYCL on Isambard.

I have an environment setup script, that takes care of bringing the corresponding libraries into the path (for XCI again).
The paths for clang and boost reference my self-built installation folders.
```bash
#! /bin/bash

source /software/arm64/spack/share/spack/setup-env.sh
module unload cmake
# load gcc & libstdc++ 11
module load gcc/11.2.0-gcc-10.3.0-module-ixozztx

export BOOST_ROOT=$HOME/install/boost

export CC=$HOME/install/llvm14/bin/clang
export CXX=$HOME/install/llvm14/bin/clang++
export PATH=$HOME/install/hipSYCL/bin/:$HOME/install/llvm14/bin/:$PATH
export LD_LIBRARY_PATH=$HOME/install/llvm14/lib/:$HOME/install/llvm14/lib/aarch64-unknown-linux-gnu/:$BOOST_ROOT/lib:$LD_LIBRARY_PATH

module load cmake/3.21.1-gcc-10.3.0-module-wuv43pg
```

On the main XCI (Thunder X2) cluster I added the following to my hipSYCL CMake line: `-DCMAKE_CXX_FLAGS=-stdlib++-isystem/software/arm64/spack/opt/spack/cray-cnl7-aarch64/gcc-10.3.0/gcc-11.2.0-ixozztxsapfg5a6ndf5kq4xwoa65uxbf/include/c++/11.2.0/ -isystem/software/arm64/spack/opt/spack/cray-cnl7-aarch64/gcc-10.3.0/gcc-11.2.0-ixozztxsapfg5a6ndf5kq4xwoa65uxbf/include/c++/11.2.0/aarch64-unknown-linux-gnu/`
This tells clang to use the libstdc++ 11 headers from that folder. Ideally, the `module load` from the env file fixes the library loading (at least I can't find any reference to me changing linking flags as well..)

For the A64FX system my configuration seems to be even more stitched together (note, I downloaded an up-to-date CMake from the website):
```bash
#! /bin/bash
export BOOST_ROOT=$HOME/a64fx/boost

export CC=$HOME/a64fx/llvm14/bin/clang
export CXX=$HOME/a64fx/llvm14/bin/clang++
export PATH=$HOME/a64fx/hipSYCL/bin:$HOME/a64fx/llvm14/bin/:$HOME/a64fx/cmake-3.22.0-rc1-linux-aarch64/bin:$PATH
export LD_LIBRARY_PATH=$HOME/a64fx/llvm14/lib/:$HOME/a64fx/llvm14/lib/aarch64-unknown-linux-gnu/:$BOOST_ROOT/lib:/software/arm64/spack/opt/spack/cray-rhel8-aarch64/gcc-8.1.0/gcc-9.3.0-lekffgwtfhjvbpj4ywzqag4mb3bjp23p/lib64/:$LD_LIBRARY_PATH
``` 

The CMake flags I used are: `-DCMAKE_CXX_FLAGS-stdlib++-isystem/software/arm64/spack/opt/spack/cray-rhel8-aarch64/gcc-8.1.0/gcc-9.3.0-lekffgwtfhjvbpj4ywzqag4mb3bjp23p/include/c++/9.3.0/ -isystem /software/arm64/spack/opt/spack/cray-rhel8-aarch64/gcc-8.1.0/gcc-9.3.0-lekffgwtfhjvbpj4ywzqag4mb3bjp23p/include/c++/9.3.0/aarch64-unknown-linux-gnu/`

Not sure whether that's all still up-to-date or whether maybe you could make it work with not building LLVM yourself and using the provided spack packages or so..

But the general rule to just get the right libstdc++ applies, `-stdlib++-isystem` and `-isystem` should be used to select the header location.

**Comment by illuhad** [comment]
Thanks @fodinabor for providing your configuration! @WageeshaR, let us know if the provided hints don't help.

I feel like we should maybe consider adding `std::tr1::filesystem` and `boost::filesystem` fallback support to hipSYCL.. While hipSYCL relying on `std::filesystem` as a C++17 code base is arguably correct, in practice this makes life difficult on a lot of HPC machines...

**Comment by WageeshaR** [comment]
Hi @fodinabor yes I'm building this on Isambard. I'm just working on a project to evaluate performance of different SYCL implementations. Thanks for sharing your config details. I will check them and update here.

**Comment by WageeshaR** [comment]
Hi sorry for the delay. @fodinabor some of your suggestions were helpful for building hipSYCL, thanks.

Now I have hipsycl built and when I try to use syclcc to compile miniFE-sycl from HeCBench, weirdly it cannot load libhipSYCL_clang.so due to boost library missing. Check the below traces.

When I `make`:
```
[ri-wshilpage@gw4a64fxlogin00 src]$ make
./get_common_files
./generate_info_header "syclcc" "-O3" "miniFE" "MINIFE"
CXX: /home/ri-wshilpage/sycl-compilers/hipSYCL/build-a64fx/install/bin/syclcc
Compiler version: clang version 11.0.0
syclcc -O3 --hipsycl-targets=omp.accelerated -I. -I../utils -I../fem -DMINIFE_SCALAR=double -DMINIFE_LOCAL_ORDINAL=int -DMINIFE_GLOBAL_ORDINAL=int -DMINIFE_RESTRICT=__restrict__ -I../../include/ -DMINIFE_CSR_MATRIX -std=c++17 -lstdc++fs  -DMINIFE_INFO=1 -DMINIFE_KERNELS=0 -c main.cpp
clang-11: warning: -lstdc++fs: 'linker' input unused [-Wunused-command-line-argument]
error: unable to load plugin '/snx11273/home/ri-wshilpage/sycl-compilers/hipSYCL/build-a64fx/install/bin/../lib/libhipSYCL_clang.so': 'libboost_fiber.so.1.73.0: cannot open shared object file: No such file or directory'
make: *** [make_targets:35: main.o] Error 1
```
But I have the above boost lib version available and it's in the ld lib path.
```
[ri-wshilpage@gw4a64fxlogin00 src]$ echo $LD_LIBRARY_PATH
/home/ri-wshilpage/sycl-compilers/hipSYCL/build-a64fx/install/lib:/lustre/projects/bristol/modules-a64fx/boost/1.73.0/arm-21.0/lib:/lustre/projects/bristol/modules-a64fx/llvm/11.0/lib:/opt/gcc/8.1.0/snos/lib64:/opt/cray/pe/lib64:/opt/cray/lib64
```
Then I checked what dynamic libs does the hipsycl lib loads,
```
[ri-wshilpage@gw4a64fxlogin00 src]$ ldd /snx11273/home/ri-wshilpage/sycl-compilers/hipSYCL/build-a64fx/install/bin/../lib/libhipSYCL_clang.so
	linux-vdso.so.1 (0x0000ffff93600000)
	libboost_fiber.so.1.73.0 => /lustre/projects/bristol/modules-a64fx/boost/1.73.0/arm-21.0/lib/libboost_fiber.so.1.73.0 (0x0000ffff934e0000)
	libboost_context.so.1.73.0 => /lustre/projects/bristol/modules-a64fx/boost/1.73.0/arm-21.0/lib/libboost_context.so.1.73.0 (0x0000ffff934c0000)
	libboost_filesystem.so.1.73.0 => /lustre/projects/bristol/modules-a64fx/boost/1.73.0/arm-21.0/lib/libboost_filesystem.so.1.73.0 (0x0000ffff93490000)
	libstdc++.so.6 => /opt/gcc/8.1.0/snos/lib64/libstdc++.so.6 (0x0000ffff932f0000)
	libm.so.6 => /lib64/libm.so.6 (0x0000ffff93200000)
	libgcc_s.so.1 => /opt/gcc/8.1.0/snos/lib64/libgcc_s.so.1 (0x0000ffff931c0000)
	libc.so.6 => /lib64/libc.so.6 (0x0000ffff93040000)
	libamath_aarch64.so => /snx11273/software/aarch64/tools/arm-compiler/21.0/arm-linux-compiler-21.0_Generic-AArch64_RHEL-8_aarch64-linux/llvm-bin/../lib/libamath_aarch64.so (0x0000ffff92eb0000)
	librt.so.1 => /lib64/librt.so.1 (0x0000ffff92e80000)
	libastring_aarch64.so => /snx11273/software/aarch64/tools/arm-compiler/21.0/arm-linux-compiler-21.0_Generic-AArch64_RHEL-8_aarch64-linux/llvm-bin/../lib/libastring_aarch64.so (0x0000ffff92e60000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x0000ffff92e20000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffff93610000)
```
Below is the Makefile of the application (only important flags):
```
CXX       = syclcc
CC        = syclcc
OPTIMIZE  = yes

CPPFLAGS = --hipsycl-targets=omp.accelerated -I. -I../utils -I../fem -I../../include/  \
           -std=c++17 -lstdc++fs

ifeq ($(OPTIMIZE),yes)
  CFLAGS += -O3
endif
```
@illuhad do you have any idea why hipSYCL_clang.so lib fails to pick boost lib? It is literally in the LD_LIBRARY_PATH and ldd shows it is picking up from the correct path. But when I try to compile the application it fails.

**Comment by illuhad** [comment]
Hmm... never seen this problem before. But I'm not sure how clang loads its plugins, and if there might be an impact on symbol resolution. Does it work if you `LD_PRELOAD` the boost library?

**Comment by illuhad** [comment]
@WageeshaR We believe that the boost dependency is no longer required for the clang plugin. Can you please try removing `${Boost_LIBRARIES}` here: https://github.com/illuhad/hipSYCL/blob/b21b73b2f6e69a37cfaee9aefb82d751ef0202ab/src/compiler/CMakeLists.txt#L70

Please let us know if this solves your issue.

**Comment by WageeshaR** [comment]
Hi, thanks. I had resolved the problem. The boost lib I was trying to load was (as I could barely believe) somewhat corrupted/mismatching. The `make` command failed when I tried `LD_PRELOAD`ing and it gave `undefined symbol: _ZTVN10__cxxabiv117__class_type_infoE` error. I thought it is working fine since it worked with `ldd` command.

However I had to switch to another boost lib and now it's working fine. I didn't check removing the dependency though. But it should work both ways.
